fondant/dataset.py

```diff
@@ -31,36 +34,74 @@ def __init__(self, manifest: Manifest):
             "__null_dask_index__": "int64",
         }
 
-    def _load_subset(self, name: str, fields: List[str]) -> dd.DataFrame:
+    def _load_subset(
+        self, name: str, fields: List[str], index: Index = None
+    ) -> dd.DataFrame:
         # get subset from the manifest
         subset = self.manifest.subsets[name]
-        # TODO remove prefix
-        location = "gcs://" + subset.location
+        # get remote path
+        remote_path = subset.location
+
+        # add index fields
+        index_fields = list(self.manifest.index.fields.keys())
+        fields = index_fields + fields
+
+        logger.info(f"Loading subset {name} with fields {fields}...")
 
         df = dd.read_parquet(
-            location,
+            remote_path,
             columns=fields,
         )
 
+        # filter on default index of manifest if no index is provided
+        if index is None:
+            index_df = self._load_index()
+            ids = index_df["id"].compute()
+            sources = index_df["source"].compute()
+            df = df[df["id"].isin(ids) & df["source"].isin(sources)]
+
+        # add subset prefix to columns
+        df = df.rename(
+            columns={
+                col: name + "_" + col for col in df.columns if col not in index_fields
+            }
+        )
+
+        return df
+
+    def _load_index(self):
+        # get index subset from the manifest
+        index = self.manifest.index
+        # get remote path
+        remote_path = index.location
+
+        df = dd.read_parquet(remote_path)
+
+        if list(df.columns) != ["id", "source"]:
+            raise ValueError(
+                f"Index columns should be 'id' and 'source', found {df.columns}"
+            )
+
         return df
 
     def load_data(self, spec: FondantComponentSpec) -> dd.DataFrame:
-        subsets = []
+        subset_dfs = []
         for name, subset in spec.input_subsets.items():
             fields = list(subset.fields.keys())
             subset_df = self._load_subset(name, fields)
-            subsets.append(subset_df)
+            subset_dfs.append(subset_df)
 
         # TODO this method should return a single dataframe with column_names called subset_field
         # TODO add index
-        # df = concatenate_datasets(subsets)
+        df = dd.concat(subset_dfs)
```

Comment:
Should we join on index rather than concatenating? Are we guaranteed that all the subsets will have aligned index and source and that they are all of the same length?

Reply:
Yes this can be addressed in a separate PR (cc @GeorgesLorre for the "return single dataframe" ticket). We might want to leverage `dd.merge` here

Reply:
alright thanks for clarifying :)